server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+    participant V as VerificationHandler
+    participant F as BlockHashingSessionFactory
+    participant S as BlockHashingSession
+    participant SV as SignatureVerifier
+    participant BSM as BlockStatusManager
+
+        
+    U->>V: (1) onBlockItemsReceived(List<BlockItem>)        
+    
+    alt (2) Detects block_header
+    
+    V->>F: createSession(initialBlockItems, executorService, signatureVerifier)
+    Note over S: New instance of BlockHashingSession created
+    F-->>V: returns BlockHashingSession (S)
+    V-->>U: Returns without blocking
+    rect rgb(230, 230, 230)
+    S->>S: *Starts hash computation asynchronously
+    end
+    
+    else (3) Append more Block Items
+    loop
+    V->>S: addBlockItems(items)
+    V-->>U: return without blocking
+    rect rgb(230, 230, 230)
+    S->>S: *Continues hash computation asynchronously
+    end
+    end
+
+    else (4) Append BlockItems with block_proof
+   
+    V->>S: addBlockItems(items with block_proof)
+    V-->>U: return without blocking    
+    S->>S: async completeHashing()   
+
+    S->>SV: (5) verifySignature(signature, computedHash, blockNumber)    
+
+    Note over SV,BSM: (6) Compare computed hash and signature
+    alt (7) Verification Fails
+      SV->>BSM: updateBlockStatus(blockNumber, ERROR, SIGNATURE_INVALID      
+    else (8) Verification Succeeds
+      SV->>BSM: updateBlockStatus(blockNumber, VERIFIED, NONE)      
+    end
+    Note over BSM: (9) Follow-up to downstream services
+    end
+
+```
+**Note:** The gray boxes on async hash computation will be implemented later only if we need to improve the latency of the verified blocks stream. (see note at the end of the document)
+## Interfaces
+
+### VerificationHandler
+
+```java
+public interface VerificationHandler {
+  void onBlockItemsReceived(List<BlockItem> blockItems);  
+}
+```
+### BlockHashingSessionFactory
+
+```java
+import java.util.concurrent.ExecutorService;
+
+public interface BlockHashingSessionFactory {
+  BlockHashingSession createSession(List<BlockItem> initialBlockItems, ExecutorService executorService, SignatureVerifier signatureVerifier);
+}
+```
+### BlockHashingSession
+```java
+/* Once hashing is completed internally, calls SignatureVerifier to continue with verification process */
+public interface BlockHashingSession {
+  void addBlockItem(BlockItem item);
+  CompletableFuture<Void> completeHashing(); // triggers final hash computation asynchronously  
+}
+```
+### SignatureVerifier
+```java
+public interface SignatureVerifier {
+  void verifySignature(byte[] signature, byte[] computedHash, long blockNumber);  
+}
+```
+### BlockStatusManager
+```java
+
+public interface BlockStatusManager {
+    void updateBlockStatus(long blockNumber, BlockVerificationStatus status, VerificationError error);
+}
+```
+
+## Enums
+```java
+public enum BlockVerificationStatus {
+  VERIFIED,
+  ERROR
+}
+
+public enum VerificationError {
+  NONE,
+  SIGNATURE_INVALID,
+  SYSTEM_ERROR
+}
+```
```

Comment:
I generally find any code at this level of design to be too much detail, but I suppose as long as we limit this to interface only, it may be OK.

Reply:
Yes, I decided to add them at the Design level to help clarify the relationships between classes. However am also wary of adding too much low level code details to a design doc, since that adds work to keep it up to date when things change on the actual code implementation. But interfaces should not change that much and is easier to maintained them so I thought that the trade-off was worth it.